mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -226,55 +245,108 @@ impl State for WaitForTakerPaymentSpend {
-            &None,
-        );
-
-        let tx = match f.compat().await {
-            Ok(t) => t,
-            Err(err) => {
-                error!("{}", err.get_plain_text_format());
-                return Self::change_state(RefundTakerPayment {});
-            },
+        let payment_search_interval = watcher_ctx.search_interval();
+        let wait_until = watcher_ctx.refund_start_time();
+        let search_input = WatcherSearchForSwapTxSpendInput {
+            time_lock: watcher_ctx.taker_locktime() as u32,
+            taker_pub: &watcher_ctx.verified_pub,
+            maker_pub: &watcher_ctx.data.maker_pub,
+            secret_hash: &watcher_ctx.data.secret_hash,
+            tx: &self.taker_payment_hex,
+            search_from_block: watcher_ctx.data.taker_coin_start_block,
         };
 
-        let tx_hash = tx.tx_hash();
-        info!("Taker payment spend tx {:02x}", tx_hash);
-        let tx_ident = TransactionIdentifier {
-            tx_hex: BytesJson::from(tx.tx_hex()),
-            tx_hash,
-        };
-
-        let secret = match watcher_ctx
-            .taker_coin
-            .extract_secret(&watcher_ctx.data.secret_hash, &tx_ident.tx_hex.0)
-            .await
-        {
-            Ok(bytes) => H256Json::from(bytes.as_slice()),
-            Err(err) => {
-                return Self::change_state(Stopped::from_reason(StopReason::Error(
-                    WatcherError::UnableToExtractSecret(err).into(),
-                )))
-            },
-        };
-
-        Self::change_state(SpendMakerPayment::new(secret))
+        loop {
+            let f = watcher_ctx
+                .taker_coin
+                .watcher_search_for_swap_tx_spend(search_input.clone())
+                .await;
+
+            match f {
+                Ok(Some(FoundSwapTxSpend::Spent(tx))) => {
+                    let now = now_ms() / 1000;
+                    if now < watcher_ctx.taker_locktime() {
+                        let wait_until = watcher_ctx.wait_for_maker_payment_spend_deadline();
+                        let maker_payment_hex_fut = watcher_ctx
+                            .maker_coin
+                            .get_tx_hex_by_hash(watcher_ctx.data.maker_payment_hash.clone());
+                        let maker_payment_hex = match maker_payment_hex_fut.compat().await {
+                            Ok(tx_res) => tx_res.tx_hex.into_vec(),
+                            Err(err) => {
+                                return Self::change_state(Stopped::from_reason(StopReason::Error(
+                                    WatcherError::MakerPaymentCouldNotBeFound(err.to_string()).into(),
+                                )))
+                            },
+                        };
+
+                        let f = watcher_ctx.maker_coin.wait_for_htlc_tx_spend(
+                            &maker_payment_hex,
+                            &watcher_ctx.data.secret_hash,
+                            wait_until,
```

Comment:
Just checking if this is the required functionality or not. Here, the watcher waits for an amount equal to taker locktime for the taker to spend the maker payment, after that, the watcher spends the maker payment for the taker if the taker didn't spend it. Any reason for this? my guess it's because of the increased fees in the transaction the watcher has but I am not sure.

Reply:
Yes that's the most important reason. Watchers don't spend the maker payment or refund the taker payment as long as there's a possibility that the taker can do it instead.